lib/auth/native/native.go

```diff
@@ -46,8 +47,79 @@ var log = logrus.WithFields(logrus.Fields{
 	trace.Component: teleport.ComponentKeyGen,
 })
 
-// PrecomputedNum is the number of keys to precompute and keep cached.
-var PrecomputedNum = 25
+// precomputedNum is the number of keys to precompute and keep cached.
+const precomputedNum = 25
```

Comment:
This whole block seems more complicated than it needs to be.

Can we use a buffered channel with capacity 25?  This would automatically queue them up and allow us to use select statements to determine when we need to generate more.

the producer side would be a loop with something like this:
```
pub, priv, err := generateKey()
// check err
select {
   case ch <- keypair{pub, priv}:
       // continue to next iteration, generate a new keypair
   case <- ctx.Done():
       return
}
```

The consumer side would just read from ch, which should have a keypair ready most of the time, and will naturally wait for the next one if the chan is empty.

Reply:
I've simplified it a bit. One of the reasons for this design was that it won't suffer from starvation if many keys are requested for a prolonged period of time. This may not be such a problem however so I switched to a design with one replenishing goroutine.

The receiver side may still generate a key in-place if none is available, the reason for this is to mitigate the above on a highly loaded auth server for example.

Reply:
I still think this would be easier to reason about with a channel. You can generate a key in place by adding a default.

```
select {
  case keypair := <- chan:
    // precomputed key available
  default:
    generateKey() // generate a key synchronously since the precompute pool is empty
}
```

Reply:
Yes, this is what the code I just pushed does?

Reply:
I must have been looking at an old diff. Looks better now.

Only remaining questions:
- do we want a context or some other way to stop the precomputation routine?
- Should we add some backoff in the case that key generation fails? I'm not sure under what circumstances it can fail, but if it does fail repeatedly, that goroutine would be in a tight loop doing compute intensive tasks.

Reply:
These are some tricky questions:
- I don't think we ever want to stop it once started, except in an error case (see below).
- This could be tricky and I'm not sure a backoff is the right solution. I just pushed an update that will stop the task upon failure and restart it during the next call to `GenerateKeyPair`. The primary failure condition here are probably only spurious but it's probably wise to guard against them anyway. This solution will result in one log entry per call and not repeated log entries due to a backoff.